### PR TITLE
[1.18.x] Make LivingSetAttackTargetEvent compatible with the Brain/Behavior system.

### PR DESCRIPTION
**Pull request by Budschie**
This PR makes the `LivingSetAttackTargetEvent` compatible with the behavior system. This allows modders to check which entities are being attacked by axolotls, zoglins, hoglins and piglins (which was not possible previously). The example mod included in this PR allows you to be ignored by piglins when holding a stick in your main hand.

The motivation of creating this PR was that I wanted players not to get attacked by piglins whenever they were morphed into one using a morph mod. Achieving this was not possible by just using events, hence I created this PR.

### CODE REVIEW

**Comment by sciwhiz12** [comment]
Our policy requires that new features or enhancements first be merged into the Latest version (1.19) before they are merged into the LTS version (1.18), unless they are non-applicable to the Latest version. 

In line with that, please create a PR or modify this PR to target 1.19 first, or leave justification on why this does not apply to 1.19.

**Comment by sciwhiz12** [comment]
For future reference, this is the LTS backport PR for #8954.

File: patches/minecraft/net/minecraft/world/entity/Mob.java.patch [hunk 1]
```diff
@@ -1,10 +1,14 @@
 --- a/net/minecraft/world/entity/Mob.java
 +++ b/net/minecraft/world/entity/Mob.java
-@@ -206,6 +_,7 @@
+@@ -206,6 +_,11 @@
  
     public void m_6710_(@Nullable LivingEntity p_21544_) {
        this.f_21362_ = p_21544_;
```
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/entity/Mob.java.patch]: Does not match merged PR. Line was removed.

File: src/main/java/net/minecraftforge/event/entity/living/LivingSetAttackTargetEvent.java [hunk 2]
```diff
@@ -7,36 +7,65 @@
...
+    
+    /**
+     * {@return the target type of this event.}
+     */
+    public ILivingTargetType getTargetType()
+    {
+        return targetType;
     }
-}
+}
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/event/entity/living/LivingSetAttackTargetEvent.java]: New line at EOF.

File: src/main/java/net/minecraftforge/event/entity/living/LivingChangeTargetEvent.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        /**
+         * This target type indicates that the target has been set by calling {@link Mob#setTarget(LivingEntity)}.
+         */
+        MOB_TARGET,
+        /**
+         * This target type indicates that the target has been set by the {@link StartAttacking} behavior.
+         */
+        BEHAVIOR_TARGET;
+    }
+}
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/event/entity/living/LivingChangeTargetEvent.java]: New line at EOF.

File: src/test/java/net/minecraftforge/debug/entity/living/LivingSetAttackTargetEventTest.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
+package net.minecraftforge.debug.entity.living;
```
Comment (ChampionAsh5357) [inline src/test/java/net/minecraftforge/debug/entity/living/LivingSetAttackTargetEventTest.java:6]: Missing license header.

File: src/test/java/net/minecraftforge/debug/entity/living/LivingSetAttackTargetEventTest.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    @SubscribeEvent
+    public void onLivingChangeTargetEvent(LivingChangeTargetEvent event)
+    {
+        // Prevents the piglin from attacking the player if they hold a stick in their hands.
+        if(event.getNewTarget() instanceof Player player && event.getEntity() instanceof AbstractPiglin piglin && player.getMainHandItem().getItem() == Items.STICK)
+        {
+            event.setCanceled(true);
+        }
+    }
+}
```
Comment (ChampionAsh5357) [inline src/test/java/net/minecraftforge/debug/entity/living/LivingSetAttackTargetEventTest.java]: New line at EOF.

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]

File: src/test/java/net/minecraftforge/debug/entity/living/LivingSetAttackTargetEventTest.java (hunk 4 above)
Comment (Budschie) [reply src/test/java/net/minecraftforge/debug/entity/living/LivingSetAttackTargetEventTest.java:6]: I actually thought that this is being added automatically. The license header might be missing in the other PR too.

File: patches/minecraft/net/minecraft/world/entity/Mob.java.patch (hunk 1 above)
Comment (Budschie) [reply patches/minecraft/net/minecraft/world/entity/Mob.java.patch]: Fixed by 8c71d59dbd73bf58841c6cf189cd307b29378e9e.

File: src/main/java/net/minecraftforge/event/entity/living/LivingSetAttackTargetEvent.java (hunk 2 above)
Comment (Budschie) [reply src/main/java/net/minecraftforge/event/entity/living/LivingSetAttackTargetEvent.java]: Fixed by c16f83fb8d8ee9619ea0d08bb7743d951591219e.

File: src/main/java/net/minecraftforge/event/entity/living/LivingChangeTargetEvent.java (hunk 3 above)
Comment (Budschie) [reply src/main/java/net/minecraftforge/event/entity/living/LivingChangeTargetEvent.java]: Fixed by c16f83fb8d8ee9619ea0d08bb7743d951591219e.

File: src/test/java/net/minecraftforge/debug/entity/living/LivingSetAttackTargetEventTest.java (hunk 4 above)
Comment (Budschie) [reply src/test/java/net/minecraftforge/debug/entity/living/LivingSetAttackTargetEventTest.java:6]: Fixed by c16f83fb8d8ee9619ea0d08bb7743d951591219e.

File: src/test/java/net/minecraftforge/debug/entity/living/LivingSetAttackTargetEventTest.java (hunk 5 above)
Comment (Budschie) [reply src/test/java/net/minecraftforge/debug/entity/living/LivingSetAttackTargetEventTest.java]: Fixed by c16f83fb8d8ee9619ea0d08bb7743d951591219e.

**Review by ChampionAsh5357 — APPROVED** [review]

File: src/test/java/net/minecraftforge/debug/entity/living/LivingSetAttackTargetEventTest.java (hunk 4 above)
Comment (TheCurle) [reply src/test/java/net/minecraftforge/debug/entity/living/LivingSetAttackTargetEventTest.java:6]: The other PR is indeed missing two licenses. Please add those here too.

Comment (TheCurle) [reply src/test/java/net/minecraftforge/debug/entity/living/LivingSetAttackTargetEventTest.java:6]: Scratch that. My brain was working on a backwards order, thought this was 1.19.

**Comment by Budschie** [comment]
Are there currently any issues with this Pull Request? I'd still be willing to resolve them. Thanks

**Comment by eerussianguy** [comment]
Its assigned, which means that its on the list of things that will get a final merge decision soon. Usually assigned PRs all get looked at at once. It's usually better to ask in discord so people see that it needs attention

**Pull request merged**